File: splearn/src/main/java/jp/kopher1601/splearn/application/required/MemberRepository.java
package jp.kopher1601.splearn.application.required;

import jp.kopher1601.splearn.domain.Email;
import jp.kopher1601.splearn.domain.Member;
import org.springframework.data.repository.Repository;

import java.util.Optional;

/**
 * 회원 정보를 저장하거나 조회한다
 */

// Repository -> Marker interface
public interface MemberRepository extends Repository<Member, Long> {
    Member save(Member member);

    Optional<Member> findByEmail(Email email);
}

File: springtx-kotlin/src/main/kotlin/jp/co/kopher1601/springtxkotlin/propagation/MemberRepository.kt
package jp.co.kopher1601.springtxkotlin.propagation

import jakarta.persistence.EntityManager
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional

@Repository
class MemberRepository(
    private val em: EntityManager,
) {

    val log = LoggerFactory.getLogger(this::class.java)

    // JPA 의 모든 데이터의 변경은 @Transactional 내에서 이루어져야 한다.
    @Transactional
    fun save(member: Member) {
        log.info("member 저장")
        em.persist(member)
    }

    fun find(username: String): Member? {
        return em.createQuery("select m from Member m where m.username = :username", Member::class.java)
            .setParameter("username", username)
            .resultList
            .firstOrNull()
    }

}